### DISCUSSION TITLE
24.11.0

### DISCUSSION THREAD

**Discussion by fballiano** [MahoCommerce/maho#58] [Announcements]
🎉🎉 **Welcome to a new release of Maho, the ecommerce platform you already know!** 🎉🎉
Remember **our last release two months ago**?  
Well, time must be an illusion since we managed to pack two years' worth of changes ([about 300 commits](https://github.com/MahoCommerce/maho/compare/24.9.0...20.11.0)) into this new one!  
For this reason we had to condense these release notes to spare you from a document that would make _Lord of the Rings_ seem short.

**🚨⚠ DISCLAIMER ⚠🚨**
Maho is a big shift in the M1 paradigm so please, for the time being, handle it with care. As much as we'd love to see production stores going live on Maho, be sure to test everything very carefully.

**Before starting, a word about the future**
We had to remove a major new feature from this release, the [global attribute editor](https://github.com/MahoCommerce/maho/pull/25), [Justin](https://github.com/justinbeaty) worked extremely hard on this (and not only on this ❤️), but the task is huge, it will bring a whole new set of features to our platform, and it simply requires a bit more time. We're confident that **it will be ready for Maho 25.01** along with [many other revolutionary changes](https://github.com/MahoCommerce/maho/discussions/52), **we aim for Maho 25.01 to have the first prototypejs-free frontend** since the beginning of the M1 platform!

This massive effort requires all the help possible - [please consider joining our sponsorship program](https://mahocommerce.com/sponsors/). It will make all the difference in the world!
[<img width="200" alt="become a sponsor" src="https://github.com/user-attachments/assets/f30438c1-0dcb-49ba-ab9c-dc8022826afb">](https://github.com/sponsors/fballiano)

Maho rocks! 🚀
https://mahocommerce.com
https://demo.mahocommerce.com

## Foundation

Maho 24.9.0 is based on OpenMage 21.0-beta2 + all the patches merged in the main branch up to 24th Nov 2024

## Visible from outside: changes to the frontend theme

<img src="https://github.com/user-attachments/assets/893fe5ba-142b-4afd-b94d-10cdb19f0d2c" width=300 alt="Maho 24.11 new frontend" align=left />

- [Completely removed jQuery](https://github.com/MahoCommerce/maho/commit/d11bb8a9d7fe6182d67b9e8ae2e216631e58d8f1),
- [All icons are now SVG](https://github.com/MahoCommerce/maho/commit/c2af7cd8a58fc426b37beaa4241d83ff9220400b)
- [Converted all font-size to rem for enhanced accessibility](https://github.com/MahoCommerce/maho/commit/72b3c3f5925a04754d54fc72232b03d41c6c545a)
- [Removed google font and replaced it with Tailwind default font rule](https://github.com/MahoCommerce/maho/commit/2d44a4f29eaf8a7f2800de89f2b9264b6998b61a)
- [Removed elevateZoom and rewrote basic zoom functionality in vanilla js](https://github.com/MahoCommerce/maho/commit/cae885fe03b9b702cfb74ffabe37fa4df329f077)
- [Removed slideshow js and rewrote them in modern CSS + basic vanilla JS](https://github.com/MahoCommerce/maho/commit/efbd58cd6c3e5b579fb0cd2503edc9d3dcf7a9d6)
- [Avoided content layout shift](https://github.com/MahoCommerce/maho/commit/a283064247f02a987beb147d58f91c7329090f7d)
- [Removed modernizr.js](https://github.com/MahoCommerce/maho/commit/208190a3453eb182cf9ca35a7e22291d130a4792)
- [Removed imagesloaded.js completely and rewrote funcionality with modern js, added lazy load to image gallery in product view](https://github.com/MahoCommerce/maho/commit/d90555d56138727339b7624cc87c33a40ba293cc)

And many, many other minor improvements.
<br clear="both"/>
<br clear="both"/>

###  [Added back/forward cache support for CMS/category/product pages](https://github.com/MahoCommerce/maho/commit/a13c68d12f5b9ca67f5f0fd9deec8720599f0db7)

<img src="https://github.com/user-attachments/assets/5918eabb-da80-48a4-831c-e222ffe607f2" width=300 align=right>

This will allow extreme performance improvements when using the back/forward buttons, better pagespeed scores, and could lead to big bandwidth/load saving on your infrastructure. ([read more](https://x.com/fballiano/status/1858170758448472197))
<br clear="both"/>

<br><br>
### [Added support for AVIF images](https://github.com/MahoCommerce/maho/commit/7993c66e1968bf67a4536b672027434a462780b4)

<img src="https://github.com/user-attachments/assets/73eb4a93-f280-4c5a-8493-9a50d6fd8ec9" width=500 align=right>

**AVIF is a modern, open-source image format** that offers superior compression efficiency while maintaining excellent image quality. It's particularly important because it **supports advanced features like HDR, wide color gamut, and transparency**, necessary to highlight the characteristics of specific type of products on your store. ([example](https://x.com/fballiano/status/1854515347770011872))
<br clear="both"/>

<br><br>
### [Added possibility to set a file format (default WebP) for all images processing (resize, thumbnails etc)](https://github.com/MahoCommerce/maho/commit/36e9e93290d9bc0a40b4db4f0c7a7eb7e53a58a9)

<img src="https://github.com/user-attachments/assets/dcf7a9f3-49e3-4aa9-b6e9-00bb54e1ca63" align=right width=300 />

Convert all your catalog to WebP with a single click. ([read more](https://x.com/fballiano/status/1857083755153813975))
<br clear="both"/>

<br><br>
### [SEO: Added automatic hreflang tag](https://github.com/MahoCommerce/maho/commit/433013a921f41cfa3fe93881f3c524d8fba017b1)

<img src="https://github.com/user-attachments/assets/9a5c75d2-8c1c-427e-9a53-ac4d5a826cb2" width=300 align=right>

This starts an effort to bring a complete set of SEO related tools to Maho. ([read more](https://x.com/fballiano/status/1856835379141452278))

From now on, **multi store website will be linked across languages** so that search engines will understand the language correlation between pages, out of the box.
<br clear="both"/>

<br><br>
### [Removed Mage_Persistent and improved cookie handling](https://github.com/MahoCommerce/maho/commit/8152a23d2451bb0c1419ee89fd6ac4f67976702e)

<img width=300 src="https://github.com/user-attachments/assets/f7bbc478-4833-4252-a555-fbd0b69accc3" align=right>

Remember the "remember me"? Well, this change removed the ancient Mage_Persistent module, consolidate and merged all templates, cleaned up a lot of code and completely rewrote the cookie management for session, you may not see the difference with your naked eye, but you will know it will work so much better.

<br clear="both"/>

<br><br>
### [Replaced js calendar by Mihai Bazon with flatpickr](https://github.com/MahoCommerce/maho/commit/d4b6e90ba71072b3e8f93b1d70a594789a3a1951) with [lazyload](https://github.com/MahoCommerce/maho/commit/20da53a9604139997fc23251f40b99cce0c2d22a)

<img src="https://github.com/user-attachments/assets/425d3f39-c16f-42da-8c9b-c91f16562ee4" width=400 align=right>

The javascript library used for dates in all M1 installation was extremely old (last release was almost 20 years ago!) and it had to go, and it's now replaced with a lightweight, moderd looking and modernly coded one called flatpickr.

<br clear="both"/>

### [Massive improvements to Maho's autoloaders](https://github.com/MahoCommerce/maho/commit/1576c673cb3f79d1aaac4800796cc256c382e239)  

The idea at the base of Maho itself was the new set of autoloaders that allow all of the platform to live inside composer's `vendor` folder, like any other modern framework would and that was never possible in the Magento world. With this PR [Justin](https://github.com/justinbeaty) took this concept and made it 10 times better.

<br><br>
### [Improved jstranslator.xml capabilities](https://github.com/MahoCommerce/maho/commit/8b8f8cbd0a7246ff86e0825a5357c9163254ca9f)

<img width="400" src="https://github.com/user-attachments/assets/252a190e-0f03-480a-b9af-81250e9fbca2" align=right>

Jstranslator gives you the possibility to add translation strings to frontend's javascript code, without having to write any javascript. It's a great feature but it needed some love. This new version **separates frontend/backend strings, allows to define a string relatively to a module, scope or if a specific javascript file is included in the target page**.  ([read more](https://github.com/MahoCommerce/maho/pull/56))
<br clear="both"/>

### [Added option to redirect to register page during checkout](https://github.com/MahoCommerce/maho/commit/2cc09e09ba77e815ff2bbc07f03000baa346387f)

When your customer registration form is too long, it's probably not the best to have it inside the checkout procedure, with this new option you can decide that your users will be redirected to the customer registration page if they choose to register during the checkout phase. Simple, but effective.

<br><br>
### [Added MahoCLI command to create a new command](https://github.com/MahoCommerce/maho/commit/44128e72e5f6d52d197a6c312621e9a492b68dcd)

<img src="https://github.com/user-attachments/assets/e161d63c-4b59-4201-b26b-47e49dba961d" width=400 align=right>

When developing an ecommerce project, you will find yourself developing some quick command line scripts, you could already integrate them into [Maho CLI](https://mahocommerce.com/cli-tool) but today this is easier than ever with `./maho create-command`.
<br clear="both"/>

<br><br>
### [Introduced `./maho health-check`](https://github.com/MahoCommerce/maho/commit/f474ab9e19fa77e1d66df6c77093a44d82e0901a) and [migrate to Maho guide](https://mahocommerce.com/migrate-to-maho)

<img src="https://github.com/user-attachments/assets/a72277cf-610f-4c4a-b36e-774c5fdf954e" width=400 align=right>

You asked us for a migration guide and we delivered, but we thought we needed something more, that's why we created `./maho health-check`, which will help you **keep your project in good health**, but also will `./maho health-check` that a project could have.

<br clear="both"/>

## ❗ Major changes you need to know
- [Products count is now disabled by default in layered navigation](https://github.com/MahoCommerce/maho/commit/08c565fe8f981798b2be38c89fcf4799738b2af8)
- [Removed CSS/JS merging, useless since HTTP2](https://github.com/MahoCommerce/maho/commit/15aa23071817bd364ad95a321a26b34a5811721b)
- [Merged openmage backend theme into the default one, which is now the only one available](https://github.com/MahoCommerce/maho/commit/48770e615c3aba73cb86564d2a9d84b568d9ae6d)
- Rewrote a huge amount of `phtml` files removing prototypejs
- [Added support for Maho-specific install scripts (#38)](https://github.com/MahoCommerce/maho/commit/737536e70df35c2de88f1546ed31258b32422f22)
- [Added JsonSerializable interface to Varien_Object](https://github.com/MahoCommerce/maho/commit/6212df9f09377a7f7c7296e88bce921b598f75c8)

## 🛡️ Security improvements
- [Fixed CodeQL warning in js/mage/adminhtml/wysiwyg/tinymce/setup.js](https://github.com/MahoCommerce/maho/commit/8f4a5641b7ac9ac99c19d80ccd9c4218867e943e)
- [Fixed CodeQL warning in js/mage/adminhtml/uploader/instance.js](https://github.com/MahoCommerce/maho/commit/d5f04ae02181b057fca5ea981b214b4b0570b03e)
- [Fixed CodeQL warning in js/mage/adminhtml/sales/packaging.js](https://github.com/MahoCommerce/maho/commit/6c8541a487098a39015e9a30670cedf07e35510b)
- [Fixed CodeQL warning in validate-url](https://github.com/MahoCommerce/maho/commit/08ce4a1ed11a9f505b237b5f52ead48571bf6ce1)
- [Fixed CodeQL warning in insertAdvice() in js/validation.js](https://github.com/MahoCommerce/maho/commit/8fae30e9ba1883b3c0131d426b35f2315e2f0313)

## 🐛 Minor bugfix, improvements and changes
- [Added new events: `controller_action_predispatch_$routeName_$controllerName` and `controller_action_postdispatch_$routeName_$controllerName`](https://github.com/MahoCommerce/maho/commit/ff4d013b4984810f54cfbee78f20ac4ddab7f540)
- [Added catalog/tags/enable_adding_tags_in_frontend configuration, disabled by default](https://github.com/MahoCommerce/maho/commit/1310b22fdd010f144f70ca3bb7316dc9093787cd)
- [API: added `maho_version` to deprecated `magentoInfo` method, also added `mahoInfo` method (#45)](https://github.com/MahoCommerce/maho/commit/81f972cbf49b7d090530f26e51438a2d66a95dc2)
- [Removed Array.From from prototype.js since it is a penalization by lighthouse and not necessary](https://github.com/MahoCommerce/maho/commit/7a9b095cace31cf3184880b1d120225f992d83b5)
- [A bug with extjs tree and missing blank image was solved](https://github.com/MahoCommerce/maho/commit/e4c49b46b57f49cb706b4b768f90397eff36a3fb)
- [Fixed a bug with Varien_Profiler and added MB information about used memory](https://github.com/MahoCommerce/maho/commit/2be00ac7e03f568ad6013ce941f6127179f720bc)
- [Minor refactor of public/api.php](https://github.com/MahoCommerce/maho/commit/8c81cdc29cd747d4d06b4315e7bdcfa49b1c4c9e)
- [Used Mage_Core_Model_App_Area constants instead of string literals](https://github.com/MahoCommerce/maho/commit/293178bf5f6a453fc779904ed8fd176889870db9)
- [Removed map files since there are no SCSS anymore](https://github.com/MahoCommerce/maho/commit/9907625d2fe1e6c4c090f60c7e23366b6d2e1d9c)
- [Removed possibility to disable "Use Web Server Rewrites" from backend](https://github.com/MahoCommerce/maho/commit/a23eb69ead434f85d5b221102ad3ace095e188f9)
- [Fixed a bug with year formatting in Mage_Customer_Block_Widget_Dob](https://github.com/MahoCommerce/maho/commit/a6c09d55fff642821bd5ceb39b38a2ed71646001)
- [Added `setTranslationHelper` method to `Mage_Core_Block_Abstract`](https://github.com/MahoCommerce/maho/commit/e759006d6b25dfaeb0e882cba3c332a2061c5391)
- [Moved minicart.js to skin/frontend/rwd/default/js since it is used only by RWD](https://github.com/MahoCommerce/maho/commit/6d30a44dd08095c99dca13f95eb703786c58f1c0)
- [Added border-collapse:collapse to "Additional Information" table in product view, for a better visual layout](https://github.com/MahoCommerce/maho/commit/063a0e4f786246437717bf8b4e313657bb311ee7)
- [Simplified product grid templates removing first/lass CSS classes and made RWD template responsive](https://github.com/MahoCommerce/maho/commit/19cbbf0e28ef7fe1e96fe919b70a51ef19495554)
- [Rewrote truncateOptions() javascript function without prototypejs](64a08d4efbdf6971b1a2a6474016e0efec6e8cc1)
- [Backend: default font is now TailWind's default font-family directive, removed OpenSans files](https://github.com/MahoCommerce/maho/commit/c904baf7f1a7f424cd101c64521f6cb37e3361c9)
- [Backend: Removed ajax-loader.gif and ajax-loader-tr.gif and migrated to loading.svg](https://github.com/MahoCommerce/maho/commit/d267aba174b74300b65f26ebd3d829610a277533)
- [Moved logo.svg and loading.svg to the skin folder of the base theme](https://github.com/MahoCommerce/maho/commit/eb361a6467fc251faa3514759f75ee94bd8afcd6)
- [Frontend: removed ajax-loader.gif and opc-ajax-loader.gif in favor of loading.svg](https://github.com/MahoCommerce/maho/commit/4e1aeb9736ebe1804a8decab6649029a16180077)
- [RWD: Removed unused class "no-js"](https://github.com/MahoCommerce/maho/commit/4dc14feee71f8a9a63a26f579c17eaaedc9e3e4c)

## 🧹 Cleanup and deprecations
- [Removed unused jscolor](https://github.com/MahoCommerce/maho/commit/3dae30f998261b3f533840940be36072e143c316)
- [Removed all "inheritDoc" from docblocks](https://github.com/MahoCommerce/maho/commit/6c05282a79df7e987215421ea0b3afeb10cc621b)
- [Directpost does not exist anymore](https://github.com/MahoCommerce/maho/commit/cbb13ff933ef0d0861d8a09807b8de90d67911e6)
- [Removed unneeded line](https://github.com/MahoCommerce/maho/commit/06faa32ea5608d4d6e9be471ec3e7986fd291450)
- [Removed deprecated controllers](https://github.com/MahoCommerce/maho/commit/4f0dc69327a98c54e2cbee79cfc02ff060e1af19)
- [PHP 8.2: Added #[\SensitiveParameter] to all the codebase](https://github.com/MahoCommerce/maho/commit/891b27d73263606d5c472d075fdec1d5c05212be)
- [Removed unused file `app/design/adminhtml/default/default/template/system/info.phtml` (#33)](https://github.com/MahoCommerce/maho/commit/e9cff1f9d2ac62de30735ff1456075e000dc548c)
- [Removed old, almost unused and undocumented lib/Unserialize](https://github.com/MahoCommerce/maho/commit/ceded0dd46bb13950b2162c4ea6a93a81787e843)
- [Removed unused Magento_Profiler (Varien_Profiler is the used one)](https://github.com/MahoCommerce/maho/commit/12016141dfd0df69f102b9e3c040beec9e1dd090)
- [Removed unused public/js/varien/menu.js](https://github.com/MahoCommerce/maho/commit/957be344a00765b6e06dbb345ddfde469479c52f)
- [Removed deprecated protected variable](https://github.com/MahoCommerce/maho/commit/f0a6d207168b3ee6753f7ae66b2bffc9d97f4c12)
- [Removed deprecated Mage_Cms_Helper_Data::isSwfDisabled()](https://github.com/MahoCommerce/maho/commit/3187e3be0d491e420c7f20e8cd6c12021a341d94)
- [Removed all ECG related lines](https://github.com/MahoCommerce/maho/commit/2a46ad20e7a2f2814381b0b6fad09a8472e41c4e)
- [Removed unused and unclear validate-clean-url](https://github.com/MahoCommerce/maho/commit/50de7d67fa4cec5a8f94030426fc2eaa5e25c435)
- [Removed unused methods from js/varien/form.js](https://github.com/MahoCommerce/maho/commit/b60119f2224600b8c6547c7f8f083ea91043346e)
- [Removed deprecated class Varien_File_Uploader_Image](https://github.com/MahoCommerce/maho/commit/7144683ef4317fe5ba3712dc1eb6003574106174)
- [Removed deprecated code from Varien_File_Uploader](https://github.com/MahoCommerce/maho/commit/ff34ba37b6f9c7f4e8d6c89f64c5001e9d1a4664)
- [Backend removed deprecated and unused hash.js](https://github.com/MahoCommerce/maho/commit/cea80e91ebcf6a4b5b5f5ae11b2664f6e1a7a627)
- [Removed all deprecated phtml files](https://github.com/MahoCommerce/maho/commit/90448cca107927f0fc8915d5f5d04012a3738304)
- [Removed unused file skin/frontend/base/default/js/giftmessage.js](https://github.com/MahoCommerce/maho/commit/365c3d75ad34d52fa63099048cd28a1ea45dd8db)
- [Removed deprecated template/page/html/top.links.phtml](https://github.com/MahoCommerce/maho/commit/17129b74e3685c51e64e91368dbd60f3efccdc5b)
- [Removed deprecated Mage_Page_Block_Html_Toplinks](https://github.com/MahoCommerce/maho/commit/2442ad53781439433475f82cc2667aad7b59f10e)
- [Removed unused skin/adminhtml/default/default/custom.css](https://github.com/MahoCommerce/maho/commit/3ee055403e88f723da94ad26dc2412149f8eb7e1)
- [Removed deprecated Mage_Adminhtml_Model_System_Config_Source_Shipping_Allowedmethods](https://github.com/MahoCommerce/maho/commit/b5c222c4fb109a20b80e0b192f4894ef49ca4b63)
- [Removed lib/Magento since almost completely unused. Migrated Magento_Db_Adapter_Pdo_Mysql to Varien_Db_Adapter_Pdo_Mysql.](https://github.com/MahoCommerce/maho/commit/fdd4ae6a3d73e21f259a75f62e93da2754d5825c)
- [RWD: removed useless font-family rules](https://github.com/MahoCommerce/maho/commit/c58b1d6922277dfe89e88ac8a2b1a602648b92eb)
- [RWD: Removed outdated default social_icons](https://github.com/MahoCommerce/maho/commit/430b2e8735b87b4b90dd0a19b289cd9723834d40)
- [Removed unused folder skin/frontend/rwd/default/scss](https://github.com/MahoCommerce/maho/commit/a3cf400b6c77cd29feff4395b422295da5631066)
- [RWD: removed ununsed file default/css/scaffold-forms.css](https://github.com/MahoCommerce/maho/commit/b166dd0d0d981158e67e3427ee25b649ccda9103)
- [RWD: removed ununsed stock pictures from default/images/media](https://github.com/MahoCommerce/maho/commit/e94480566b4bf7276aa17972b0adb08ae8ca022a)

## 🛠️ Technical updates
- All dependencies are updated to the last available version
- [Updated to PHPStan 2.0.2](https://github.com/MahoCommerce/maho/commit/1ec4f1a75447d2598c60ececf0b05251df46a0e1)
- [PHPStan: added result cache](https://github.com/MahoCommerce/maho/commit/1891f88467745938b90b94fa741dc645d509ef54)
- [Moved rector to dev-main waiting for official support for PHPStan2](https://github.com/MahoCommerce/maho/commit/e3a987f8e560db09fa96f0080b079a70904aa433)
- [Synced required PHP extensions from composer.json to Mage/Install/etc/install.xml](https://github.com/MahoCommerce/maho/commit/83dbab4d5626b2d75e75ea89adbe7220a5da6b14)
- [Updated CodeQL configuration reflecting new project structure](https://github.com/MahoCommerce/maho/commit/38f80f675b0ed902492ed728929294d10a7e1ab9)
- [Added workflow to avoid committing \r line ending](https://github.com/MahoCommerce/maho/commit/b6639c9283eade246ed37ae3afe8577233393d41)
- [Bumped pelago/emogrifier 7.3.0](https://github.com/MahoCommerce/maho/commit/68507ea3d0af785c8e6b8e486f1c1fd2eb0c0eaf)
- [Bumped htmlpurifier to 4.18.0](https://github.com/MahoCommerce/maho/commit/09c4c35b48f638c15ee606178ff7763fc58daa52)
- [Better configuration for PHPCS](https://github.com/MahoCommerce/maho/commit/08eccbc8a300a0607f51db8eb9ec024bbc94ef1e)
- [Replaced CRLF with LF line separators](https://github.com/MahoCommerce/maho/commit/7e93ad1fcc8e0b3b8ad12d2aea5b82b7fa332b12)
- [Moved composer patches to Maho repository](https://github.com/MahoCommerce/maho/commit/73014072b0b52d20d7d573d8ce3c9371ccb4fae2)

**Full Changelog**: https://github.com/MahoCommerce/maho/compare/24.9.0...24.11.0

<hr /><em>This discussion was created from the release <a href='https://github.com/MahoCommerce/maho/releases/tag/24.11.0'>24.11.0</a>.</em>

**Comment by toejam34** [comment]
Impressive work team. 

Is there a build version that you are targeting where you might say that now is a good time to move from m1 ( or openmage ) to Maho?

**Reply by fballiano** [reply]
Thanks @toejam34, since this version includes a huge work on the autoloaders (fixing many first-generation problems that were present in 24.9) I'd say it could be worth a test. Migration really depends on how complex your projects are but... every release will have its challenges because we want to add/change a lot of stuff :-)

**Reply by toejam34** [reply]
The site I look after is pretty vanilla with a few addons that appear to work on openmage with php 8.3. The only issue I think I might encounter is with a worldpay plugin. 

I guess I could create a development site and test thoroughly on there. 

Also I suspect once I am in the Maho eco system then the steps between builds should be easier to manage.

**Reply by fballiano** [reply]
it would be great to have some real world test 😍

**Reply by justinbeaty** [reply]
One caveat is I think you may need to manually copy any module's `js` and `skin` directories to your public folder for now.
